### PR TITLE
Infra/picker testing coverage

### PR DESCRIPTION
**Pull request by adids1221**
## Description
Picker testing coverage.
Testing picker modes (modal and dialog) - currently there is an issue with WheelPicker testing.
Testing picker field types by style.
Search functionality testing have the same issue as the WheelPicker.

## Changelog
Picker testing coverage

## Additional info
MADS-4181

### CODE REVIEW

**Review by M-i-k-e-l — APPROVED** [review]

File: src/components/picker/Picker.driver.new.ts [hunk 1]
```diff
@@ -60,13 +60,18 @@ export const PickerDriver = (props: ComponentProps) => {
     itemDriver.press();
   };
 
+  const getInput = () => {
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply src/components/picker/Picker.driver.new.ts:63]: TBH I feel like this should not be exposed as a driver (do you need anything except the `exists`?)

File: src/components/picker/__tests__/index.spec.tsx [hunk 2]
```diff
@@ -100,122 +97,185 @@ describe('Picker', () => {
...
+      screen.debug();
+      const textFieldDriver = driver.getInput();
+      act(() => expect(textFieldDriver.exists()).toBeTruthy());
+      expect(textFieldDriver.getStyle()).toEqual(expect.objectContaining(Typography.text70));
+    });
+    it('should render a settings picker', async () => {
+      const driver = getDriver({fieldType: 'settings'});
+      const textFieldDriver = driver.getInput();
+      act(() => expect(textFieldDriver.exists()).toBeTruthy());
+      console.log(`textFieldDriver.getStyle():`, textFieldDriver.getStyle());
```
Comment (M-i-k-e-l) [inline src/components/picker/__tests__/index.spec.tsx:277]: Reminder to delete

File: src/components/picker/Picker.driver.new.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply src/components/picker/Picker.driver.new.ts:63]: Maybe there needs to be a different `testID` for each field type

File: src/components/picker/__tests__/index.spec.tsx [hunk 3]
```diff
@@ -100,122 +97,185 @@ describe('Picker', () => {
...
+      // });
+
+      describe('Test Modal TopBar', () => {
+        it('should close and select items when pressing on cancel button', () => {
+          const driver = getDriver({mode: 'MULTI', topBarProps: {cancelLabel: 'Cancel'}});
+          driver.open();
+          expect(driver.isOpen()).toBeTruthy();
+          driver.selectItem(countries[2].label);
+          driver.selectItem(countries[4].label);
+          driver.done();
```
Comment (M-i-k-e-l) [inline src/components/picker/__tests__/index.spec.tsx:143]: You're passing `cancel` button and the test is saying `cancel` and here you're pressing `done`, I'm confused

File: src/components/picker/__tests__/index.spec.tsx [hunk 4]
```diff
@@ -100,122 +97,185 @@ describe('Picker', () => {
...
   // describe.skip('WheelPicker', () => {
   // });
+
+  //TODO: add more tests for different props
+  describe('Picker field types', () => {
+    it('should render a form picker', () => {
+      const driver = getDriver({fieldType: 'form'});
+      const textFieldDriver = driver.getInput();
+      act(() => expect(textFieldDriver.exists()).toBeTruthy());
+      expect(textFieldDriver.getStyle()).toEqual(expect.objectContaining({textAlign: 'left'}));
```
Comment (M-i-k-e-l) [inline src/components/picker/__tests__/index.spec.tsx:264]: What if the design changes?
Maybe the `testID`s should be different and verify this way; not sure.

File: src/components/picker/__tests__/index.spec.tsx [hunk 5]
```diff
@@ -100,122 +97,185 @@ describe('Picker', () => {
...
+    });
+    it('should render a filter picker', () => {
+      const driver = getDriver({fieldType: 'filter'});
+      screen.debug();
+      const textFieldDriver = driver.getInput();
+      act(() => expect(textFieldDriver.exists()).toBeTruthy());
+      expect(textFieldDriver.getStyle()).toEqual(expect.objectContaining(Typography.text70));
```
Comment (M-i-k-e-l) [inline src/components/picker/__tests__/index.spec.tsx:271]: What if the design changes?
Maybe the `testID`s should be different and verify this way; not sure.

File: src/components/picker/__tests__/index.spec.tsx [hunk 6]
```diff
@@ -100,122 +97,185 @@ describe('Picker', () => {
...
+    });
+    it('should render a settings picker', async () => {
+      const driver = getDriver({fieldType: 'settings'});
+      const textFieldDriver = driver.getInput();
+      act(() => expect(textFieldDriver.exists()).toBeTruthy());
+      console.log(`textFieldDriver.getStyle():`, textFieldDriver.getStyle());
+      expect(textFieldDriver.getStyle()).toEqual(expect.objectContaining(Typography.text70));
```
Comment (M-i-k-e-l) [inline src/components/picker/__tests__/index.spec.tsx:278]: What if the design changes?
Maybe the `testID`s should be different and verify this way; not sure.

File: src/components/picker/__tests__/index.spec.tsx [hunk 7]
```diff
@@ -100,122 +97,185 @@ describe('Picker', () => {
...
         expect(driver.isOpen()).toBeFalsy();
-        // driver.open();
-        // expect(driver.isOpen()).toBeTruthy();
         jest.useFakeTimers();
         act(() => driver.open());
         jest.runAllTimers();
-        // advanceAnimationByTime(10000);
-        // await new Promise(r => setTimeout(r, 1000));
-        await waitFor(() => expect(driver.isOpen()).toBeTruthy());
+        await waitFor(() => expect(driver.isOpen(dialogProps.useDialog)).toBeTruthy());
```
Comment (M-i-k-e-l) [inline src/components/picker/__tests__/index.spec.tsx:189]: I don't like the fact that we need to send the prop `useDialog` here.
We can take it to the other devs.

File: src/components/picker/Picker.driver.new.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (adids1221) [reply src/components/picker/Picker.driver.new.ts:63]: I added relevant testID to the relevant fields, tested `Settings and Filter` field types.

File: src/components/picker/__tests__/index.spec.tsx (hunk 7 above)
Comment (adids1221) [reply src/components/picker/__tests__/index.spec.tsx:189]: Added to the next dev sync.
I'll ask Ethan about it tomorrow in our private sync.

File: src/components/picker/__tests__/index.spec.tsx [hunk 8]
```diff
@@ -100,122 +96,191 @@ describe('Picker', () => {
...
   // });
+
+  //TODO: add more tests for different props
+  describe('Picker field types', () => {
+    describe('Test filter field type', () => {
+      const placeholderText = 'Select a Filter';
+      it('should render a filter picker', () => {
+        const driver = getDriver({fieldType: 'filter', placeholder: placeholderText});
+        expect(driver.isOpen()).toBeFalsy();
+        screen.debug();
```
Comment (M-i-k-e-l) [inline src/components/picker/__tests__/index.spec.tsx:265]: Is this here on purpose?

File: src/components/picker/__tests__/index.spec.tsx [hunk 9]
```diff
@@ -100,122 +96,191 @@ describe('Picker', () => {
...
+          }
+        });
+        driver.open();
+        expect(driver.isOpen()).toBeTruthy();
+        driver.cancel();
+        expect(driver.isOpen()).toBeFalsy();
+        expect(onDismiss).toHaveBeenCalledTimes(2); // TODO: this should be 1
+      });
+
+      // TODO: this test is not passing yet
```
Comment (M-i-k-e-l) [inline src/components/picker/__tests__/index.spec.tsx:122]: When should this start working?

File: src/components/picker/__tests__/index.spec.tsx [hunk 10]
```diff
@@ -100,122 +96,191 @@ describe('Picker', () => {
...
-      expect(driver.isOpen()).toBeFalsy();
+    it('Test onPress', () => {
+      const driver = getDriver({onPress});
       driver.open();
       expect(driver.isOpen()).toBeTruthy();
       driver.cancel();
       expect(driver.isOpen()).toBeFalsy();
-      expect(onDismiss).toHaveBeenCalledTimes(2); // TODO: this should be 1
+      expect(onPress).toHaveBeenCalled();
+    });
```
Comment (M-i-k-e-l) [inline src/components/picker/__tests__/index.spec.tsx:106]: Shouldn't this be only this:
```
const driver = getDriver({onPress});
driver.open();
expect(onPress).toHaveBeenCalled();
```

File: src/components/picker/__tests__/index.spec.tsx [hunk 11]
```diff
@@ -100,122 +96,191 @@ describe('Picker', () => {
...
-        // await new Promise(r => setTimeout(r, 1000));
         await waitFor(() => expect(driver.isOpen()).toBeTruthy());
       });
 
-      it('Should not open when disabled', () => {
-        const driver = getDriver({useDialog: true, customPickerProps: {migrateDialog: true}, editable: false});
-
-        expect(driver.isOpen()).toBeFalsy();
-        driver.open();
+      it('Should not open when disabled', async () => {
```
Comment (M-i-k-e-l) [inline src/components/picker/__tests__/index.spec.tsx:192]: Looks like we no longer need the `async` (`await`, `act` and `waitFor`)

File: src/components/picker/__tests__/index.spec.tsx [hunk 12]
```diff
@@ -100,122 +96,191 @@ describe('Picker', () => {
...
   });
 
-  // TODO: add tests for WheelPicker as well
+  // TODO: add tests for WheelPicker
+  // Note: the picker dialog should be opened and then the wheel picker should be rendered, this is the main issue with testing it for now
   // describe.skip('WheelPicker', () => {
   // });
+
+  //TODO: add more tests for different props
+  describe('Picker field types', () => {
```
Comment (M-i-k-e-l) [inline src/components/picker/__tests__/index.spec.tsx:259]: These are failing

File: src/components/textField/TextField.driver.new.ts [hunk 13]
```diff
@@ -79,6 +80,10 @@ export const TextFieldDriver = (props: ComponentProps) => {
     return charCounterDriver;
   };
 
+  const getStyle = () => {
+    return StyleSheet.flatten(driver.getElement().props.style) as TextStyle;
+  };
```
Comment (M-i-k-e-l) [inline src/components/textField/TextField.driver.new.ts:85]: I don't think you're using this anymore, do we want to keep this change?

File: src/components/picker/__tests__/index.spec.tsx (hunk 8 above)
Comment (adids1221) [reply src/components/picker/__tests__/index.spec.tsx:265]: no, I'll delete it

File: src/components/picker/__tests__/index.spec.tsx (hunk 9 above)
Comment (adids1221) [reply src/components/picker/__tests__/index.spec.tsx:122]: The issue here is that the `onShow` function is triggered only when the modal is fully open.
Iv'e tried to use `act` and `waitFor` didn't help.

I'll change the `TODO` comment to fix this test in the future, WDYT ?

File: src/components/picker/__tests__/index.spec.tsx (hunk 10 above)
Comment (adids1221) [reply src/components/picker/__tests__/index.spec.tsx:106]: you are right

File: src/components/picker/__tests__/index.spec.tsx (hunk 12 above)
Comment (adids1221) [reply src/components/picker/__tests__/index.spec.tsx:259]: fixed, bad merging from master after merging the new `useFieldType` hook.

File: src/components/picker/__tests__/index.spec.tsx (hunk 8 above)
Comment (adids1221) [reply src/components/picker/__tests__/index.spec.tsx]: removed

File: src/components/picker/__tests__/index.spec.tsx (hunk 9 above)
Comment (M-i-k-e-l) [reply src/components/picker/__tests__/index.spec.tsx:122]: Doesn't matter IMO

File: src/components/picker/__tests__/index.spec.tsx (hunk 10 above)
Comment (M-i-k-e-l) [reply src/components/picker/__tests__/index.spec.tsx:106]: I don't see a fix for this; IMO it should be:
```
it('Test onPress', () => {
  const driver = getDriver({onPress});
  driver.open();
  expect(onPress).toHaveBeenCalled();
});

```

**Review by M-i-k-e-l — COMMENTED** [review]
@adids1221  Are you sure you've pushed your changed?

**Review by M-i-k-e-l — APPROVED** [review]
Approved, please see the last note about the driver

File: src/components/picker/__tests__/index.spec.tsx [hunk 14]
```diff
@@ -100,122 +96,199 @@ describe('Picker', () => {
...
+
+      it('should render a settings picker with placeholder', async () => {
+        const driver = getDriver({fieldType: 'settings', placeholder: placeholderText});
+        expect(driver.isOpen()).toBeFalsy();
+        const label = screen.getByTestId(`${testID}.settings.type.label`);
+        const placeholder = screen.getByTestId(`${testID}.settings.type.placeholder`);
+        expect(label).toBeTruthy();
+        expect(placeholder).toBeTruthy();
+        expect(label.props.children).toEqual(undefined);
+        expect(placeholder.props.children).toEqual(placeholderText);
```
[the remark this answers is not in the document]
Comment (adids1221) [reply src/components/picker/__tests__/index.spec.tsx:290]: Ticket MADS-4318

**Pull request merged**